File: 2016/Day_03/Ruby/solution.rb
require "../../input_parsing/parse_input"

def is_a_triangle(side_1, side_2, side_3)
  if side_1 + side_2 > side_3 and side_2 + side_3 >  side_1 and side_1 + side_3 > side_2
    1
  else
    0
  end
end

triangle_triples_input = input_per_line("../input.txt")
triangle_count_part_1 = 0
triangle_count_part_2 = 0
triangle_part_two_array = []

triangle_triples_input.each do |triple|
  this_triple = triple.split(" ")
  triangle_count_part_1 += is_a_triangle(this_triple[0].to_i, this_triple[1].to_i, this_triple[2].to_i)
end

puts "If we are looking at rows of triplets, there are #{triangle_count_part_1} triangles"

File: 2016/Day_03/Perl/solution.pl
#! /usr/bin/perl

use v5.20;

open (TRIANGLES, "../input.txt") || die "No triangles!";

my @potential_triangles = <TRIANGLES>;

sub is_it_a_triangle{

    my $side_1 = $_[0];
    my $side_2 = $_[1];
    my $side_3 = $_[2];
    
    if (($side_1 + $side_2 > $side_3) && ($side_2 + $side_3 > $side_1) && ($side_1 + $side_3 > $side_2)){
        return 1;
    }
    return 0;
}

my $part_one_count = 0;
my $part_two_count = 0;
my @part_2_triangle_1;
my @part_2_triangle_2;
my @part_2_triangle_3;
my $part_two_iterator = 0;

foreach my $triple (@potential_triangles){
    my @useful_triple = split(" ", $triple);
    $part_one_count += is_it_a_triangle($useful_triple[0], $useful_triple[1], $useful_triple[2]);
    $part_two_iterator += 1;
    push @part_2_triangle_1, $useful_triple[0];
    push @part_2_triangle_2, $useful_triple[1];
    push @part_2_triangle_3, $useful_triple[2];
    if ($part_two_iterator % 3 == 0){
        $part_two_count += is_it_a_triangle(pop @part_2_triangle_1, pop @part_2_triangle_1, pop @part_2_triangle_1);
        $part_two_count += is_it_a_triangle(pop @part_2_triangle_2, pop @part_2_triangle_2, pop @part_2_triangle_2);
        $part_two_count += is_it_a_triangle(pop @part_2_triangle_3, pop @part_2_triangle_3, pop @part_2_triangle_3);
    }
}

say "If we are looking across rows, we have $part_one_count triangles.";
say "If we are looking across columns, we have $part_two_count triangles.";
